File: Assistant/lib/view/modding_worker/forward_helper.dart
import '/common.dart';
import '/module.dart';
import '/setting.dart';
import '/utility/json_helper.dart';
import 'package:flutter/widgets.dart';
import 'package:provider/provider.dart';

// ----------------

class ForwardHelper {

  // #region utility

  static Future<Void> forward(
    BuildContext context,
    List<String> argument,
  ) async {
    var setting = Provider.of<SettingProvider>(context, listen: false);
    await setting.state.handleLaunch!(
      ModuleHelper.query(ModuleType.modding_worker).name,
      ModuleType.modding_worker,
      ['-additional_argument', ...argument],
    );
    return;
  }

  static Future<Void> forwardMany(
    BuildContext       context,
    List<List<String>> argument,
    Boolean            parallel,
  ) async {
    if (!parallel) {
      await forward(context, argument.expand((value) => [...value]).toList());
    }
    else {
      for (var argumentItem in argument) {
        await forward(context, argumentItem);
      }
    }
    return;
  }

  // ----------------

  static String makeMethodMaybeBatch(
    String  method,
    Boolean batch,
  ) {
    return '${method}${!batch ? '' : '!batch'}';
  }

  static List<String> makeArgumentForCommand(
    String? input,
    String? method,
    Object? argument,
  ) {
    var command = <String>[];
    command.add(input ?? '?none');
    if (method != null) {
      command.add('-method');
      command.add(method);
    }
    if (argument != null) {
      command.add('-argument');
      command.add(JsonHelper.serializeText(argument, indented: false));
    }
    return command;
  }

  // #endregion

}

File: AssistantPlus/Application/View/CoreModdingWorker/ForwardHelper.cs
#pragma warning disable 0,
// ReSharper disable

using Twinning.AssistantPlus;
using Twinning.AssistantPlus.Utility;

namespace Twinning.AssistantPlus.View.CoreModdingWorker {

	public static class ForwardHelper {

		#region utility

		public static async Task Forward (
			List<String> argument
		) {
			await App.Instance.HandleLaunch(
				ModuleHelper.Query(ModuleType.CoreModdingWorker).Name,
				ModuleType.CoreModdingWorker,
				["-additional_argument", ..argument]
			);
			return;
		}

		public static async Task ForwardMany (
			List<List<String>> argument,
			Boolean            parallel
		) {
			if (!parallel) {
				await ForwardHelper.Forward(argument.SelectMany((value) => (value)).ToList());
			}
			else {
				foreach (var argumentItem in argument) {
					await ForwardHelper.Forward(argumentItem);
				}
			}
			return;
		}

		// ----------------

		public static String MakeMethodMaybeBatch (
			String  method,
			Boolean batch
		) {
			return $"{method}{(!batch ? "" : "!batch")}";
		}

		public static List<String> MakeArgumentForCommand (
			String?                     input,
			String?                     method,
			Dictionary<String, Object>? argument
		) {
			var command = new List<String>();
			command.Add(input ?? "?none");
			if (method != null) {
				command.Add("-method");
				command.Add(method);
			}
			if (argument != null) {
				command.Add("-argument");
				command.Add(JsonHelper.SerializeText(argument, false));
			}
			return command;
		}

		#endregion

	}

}
